XSS Prevention Cheatsheet Rewrite v2

colecornford opened the pull request:
Hello friends,

Round 2 for PR, have taken onboard all Jakub and Jim's feedback to produce this version. Happy to take back and fix whatever is missing.

Cole Cornford

Thank you for submitting a Pull Request (PR) to the Cheat Sheet Series. 

> :triangular_flag_on_post: If your PR is related to grammar/typo mistakes, please double-check the file for other mistakes in order to fix all the issues in the current cheat sheet.

Please make sure that for your contribution:

- [ ] In case of a new Cheat Sheet, you have used the [Cheat Sheet template](https://github.com/OWASP/CheatSheetSeries/blob/master/templates/New_CheatSheet.md).
- [x] All the markdown files do not raise any validation policy violation, see the [policy](https://github.com/OWASP/CheatSheetSeries/actions?query=workflow%3A%22Markdown+Link+Check%22).
- [x] All the markdown files follow these [format rules](https://github.com/OWASP/CheatSheetSeries/blob/master/CONTRIBUTING.md#markdown).
- [x] All your assets are stored in the **assets** folder.
- [x] All the images used are in the **PNG** format.
- [x] Any references to websites have been formatted as [TEXT](URL)
- [x] You verified/tested the effectiveness of your contribution (e.g., the defensive code proposed is really an effective remediation? Please verify it works!).
- [ ] The CI build of your PR pass, see the build status [here](https://github.com/OWASP/CheatSheetSeries/actions).

If your PR is related to an issue, please finish your PR text with the following line:

This PR covers issue #<insert number here>.

Thank you again for your contribution :smiley:

ThunderSon commented on cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.md:
```diff
@@ -1,400 +1,207 @@
-# Cross Site Scripting Prevention Cheat Sheet
+# Cross Site Scripting (XSS) Prevention Cheat Sheet
```
AFAIK having the shorter version in the title isn't a great practice, and having it defined in the first sentence is better.
Makes linking as well easier on processors (sometimes they break on special chars, `(` and `)` in this case)

mackowski replied:
I agree with @ThunderSon

colecornford replied:
Accepted.

ThunderSon commented on cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.md:
```diff
@@ -1,400 +1,207 @@
+Cross-Site Scripting is a misnomer. The name originated from early versions of the attack where stealing data cross-site was the primary focus. Since then, it has extended to include injection of basically any content, but we still refer to this as XSS. 
 
-Both [reflected and stored XSS](https://owasp.org/www-community/attacks/xss/#stored-and-reflected-xss-attacks) can be addressed by performing the appropriate validation and encoding on the server-side. [DOM Based XSS](https://owasp.org/www-community/attacks/DOM_Based_XSS) can be addressed with a special subset of rules described in the [DOM based XSS Prevention Cheat Sheet](DOM_based_XSS_Preventi ...
+XSS is serious and can lead to account impersonation, taking actions as a user, observing user behaviour, loading external content, and stealing sensitive data. This cheatsheet is a list of techniques to prevent or limit the impact of XSS. No single technique will solve XSS. Using the right combination of Framework Security Measures, Output Encoding, and HTML Sanitization will prevent most XSS at ...
```
```suggestion
XSS is serious and can lead to account impersonation, taking actions as a user, observing user behaviour, loading external content, and stealing sensitive data. This cheatsheet is a list of techniques to prevent or limit the impact of XSS. No single technique will solve XSS. Using the right combination of Framework Security Measures, Output Encoding, and HTML Sanitization techniques will prevent most XSS attacks.
```

mackowski commented:
@colecornford thanks for this - can you please check some lintech errors and fix them?

mackowski commented:
@jmanico can you please review this first as I know that you have a strong opinions on this :) 
I remember that we agreed on refreshing this but also keeping some of the sections from old CS https://github.com/OWASP/CheatSheetSeries/issues/517

ThunderSon reviewed (commented):
I did a first review of this, as a new CS, without comparing it with the previous one.
I'll give this a look later on :)

Great job overall!

colecornford commented on cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.md:
```diff
@@ -1,400 +1,207 @@
+XSS is serious and can lead to account impersonation, taking actions as a user, observing user behaviour, loading external content, and stealing sensitive data. This cheatsheet is a list of techniques to prevent or limit the impact of XSS. No single technique will solve XSS. Using the right combination of Framework Security Measures, Output Encoding, and HTML Sanitization will prevent most XSS at ...
 
-For a cheatsheet on the attack vectors related to XSS, please refer to the [XSS Filter Evasion Cheat Sheet](https://owasp.org/www-community/xss-filter-evasion-cheatsheet). More background on browser security and the various browsers can be found in the [Browser Security Handbook](https://code.google.com/archive/p/browsersec/).
+## 2022 Updates
```
Accepted, will move into PR discussion.

colecornford commented on cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.md:
```diff
@@ -1,400 +1,207 @@
+## XSS Defense Philosophy
 
-### You Need a Security Encoding Library
+For XSS attacks to be successful, an attacker needs to insert and execute malicious content in a webpage. We want to stop this by protecting every variable in your web application. All variables must go through validation, escaping, and sanitization. We call this perfect injection resistance. If any variable gets a free ride, you’re introducing a weakness that may become an XSS vulnerability.
```
Agreed with point A) will accept and change.

Will remove Royal "we". I missed that aspect as I was writing this part w/ Jim word for word.

colecornford commented on cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.md:
```diff
@@ -1,400 +1,207 @@
+For XSS attacks to be successful, an attacker needs to insert and execute malicious content in a webpage. We want to stop this by protecting every variable in your web application. All variables must go through validation, escaping, and sanitization. We call this perfect injection resistance. If any variable gets a free ride, you’re introducing a weakness that may become an XSS vulnerability.
 
-Writing these encoders is not tremendously difficult, but there are quite a few hidden pitfalls. For example, you might be tempted to use some of the escaping shortcuts like `\"` in JavaScript. However, these values are dangerous and may be misinterpreted by the nested parsers in the browser. You might also forget to escape the escape character, which attackers can use to neutralize your attempts ...
+Frameworks have dramatically reduced the areas where variables miss one of these steps. But they aren't perfect, and that is why we need to continue to use Output Encoding and HTML Sanitization.
```
Yep will edit further. Accepted.

colecornford commented on cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.md:
```diff
@@ -1,400 +1,207 @@
+Output Encoding is recommended when you need to safely display data exactly as a user typed it in. We do not want variables to be interpreted as code instead of text. This section covers each form of output encoding, where to use it, and where to avoid using dynamic variables entirely.
 
-## XSS Prevention Rules
+Start with using your framework’s default protections. Automatic encoding and escaping functions are built into most frameworks. For further details, check out our [React](), [Vue](), and [Angular]() guides (available April 2022).
```
Will remove hyperlinks and add at a later date once created. Accepted.

colecornford commented on cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.md:
Yes so the thought process for this was "I am heavily editing 40 pages of content into 10". 🤣 

So, the advice here is from the prior version of the cheatsheet and while it might still be accurate, I was planning on doing a second pass to review the currency and applicability of the technical aspects after the overall rework. Hope that clarifies why the `\xHH` recommendation is there for now.

colecornford commented on cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.md:
Yep as prior comment, I wanted to start with cleaning content before moving to updating it to be current and easily applicable. I think verifying content-type alone is okay as JSON is usually expected to be interpreted as... json.. lol

colecornford commented on cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.md:
Might be worthwhile for us to identify which libraries exist for doing Output Encoding across major languages and putting them into an appendix.

colecornford commented:
Hi friends, 
@mackowski > All linting passed and addressed using _ instead of **
@ThunderSon > Addressed all feedback. I believe we do need to update the core advice specifically surrounding "Encode Everything" to be more lenient if folk are using quotes. The core issue from the previous CheatSheet came from folks forgetting to close HTML attributes and similar, resulting in many characters being able to be used for escaping into a JavaScript context. So I added a section saying "Please make sure you quote things". But I am open to discuss w/ @jmanico and @mackowski whether we want to simplify the language under each section to characters relevant for that context IF strict quoting is followed.

Anyway, I'm going to do some work on ZAP while you have another look through my updates :D

colecornford commented:
@jmanico > Most feedback has been addressed above. Thoughts on less-restrictive character set for output encoding if within quoted attributes?

jmanico commented:
I do not want to over-write the old work yet. Can we please deliver this as a draft and when this process is done, do a change over?

mackowski commented:
@jmanico i understand that you want to add this to the drafts to make some more changes before we do on prod. Can we make these changes in this PR? Or will it be easier to merge it into drafts and make more PR?

jmanico commented:
Right now the work is written in an informal language. I want to do one big edit pass and convert it to more professional language. So my suggestion is - first complete this work as a draft somehow, let me do my last round of edit, and then let's socialize this and let folks review new and old cheatsheet before we push this live. We're overwriting one of the most popular OWASP pages in history, let's tred lightly and take these steps carefully.

Or, please commit this with a different name that does not overwrite the old cheatsheet.

This is just my request, not a demand. Is this acceptable?

jmanico commented:
Also, the link checker is failing. :)

jmanico commented:
> 



> @jmanico > Most feedback has been addressed above. Thoughts on less-restrictive character set for output encoding if within quoted attributes?

I'm ok to just mention HTML entity encoding for both HTML and Attribute contexts to simplify. Your call. :) We'll have a massive public review and get lots more feedback before we go live.

mackowski commented:
@jmanico good idea to have in drafts and run a second round of review :) 
@colecornford can you fix the lint errors and move this cheatsheet to cheatsheets_draft (leaving the old one)

colecornford commented:
Reverted original xss cheatsheet, moved new CS into draft folder, fixed lint issues surrounding acronym casing + apostrophe usage.

mackowski commented:
Thanks @colecornford you are awesome! I just run the linter :)

colecornford commented:
Hey @mackowski I found the linter issue... I was missing one whitespace character. Fixed.

colecornford commented:
Hey @jmanico > Fixed broken link from the table at the end from prior cheatsheet, removed all 'Cute' language. It's serious now.
